### ISSUE TITLE
CEL client side validation

### ISSUE THREAD

**Issue by clux** [kube-rs/kube#1670]
### Would you like to work on this feature?

no

### What problem are you trying to solve?

I want to verify that a local object satisfies a CEL validation rule as specified via `Rule` or `#[cel_validation]` attributes (via https://github.com/kube-rs/kube/pull/1649 ) without having to `kubectl apply` / `Api::patch` the object and wait for apiserver validation to pass/fail.

### Describe the solution you'd like

An optional feature dependency `kube-core/cel` and maybe `kube-derive/cel` (enabled via `kube/cel`) which adds a dependency on (e.g.) [cel-interpreter](https://crates.io/crates/cel-interpreter) to run client-side CEL evaluation against local objects.

AFAIKT, we can encode our objects as a cel [Context](https://docs.rs/cel-interpreter/0.9.0/cel_interpreter/context/enum.Context.html) and run the `x-kubernetes-validations.rule` against it. Their [`Value`](https://docs.rs/cel-interpreter/latest/cel_interpreter/objects/enum.Value.html) type seems to be convertible into [from a serde::Serialize impl](https://docs.rs/cel-interpreter/latest/cel_interpreter/fn.to_value.html).

### Describe alternatives you've considered

None, open to alternatives.

### Documentation, Adoption, Migration Strategy

Addition under feature for `CELSchema` users. Needs documentation as with the rest of cel.

### Target crate for feature

kube-core, kube-derive

**Comment by doxxx93** [comment]
I've been working on [kube-cel](https://github.com/doxxx93/kube-cel) ([crates.io](https://crates.io/crates/kube-cel)), a standalone crate that implements the client-side CEL validation pipeline described here. Built on the [`cel`](https://crates.io/crates/cel) crate (the renamed `cel-interpreter` mentioned in this issue).

What it does today (v0.4):

- Kubernetes CEL extension functions (strings, lists, sets, regex, urls, ip/cidr, semver, quantity, format, math, base64, jsonpatch)
- Math library — `math.ceil`, `math.floor`, `math.round`, `math.abs`, `math.sign`, `math.sqrt`, bitwise ops, `math.greatest`/`math.least` (matching cel-go/ext)
- Named format validation — 13 format validators (`format.dns1123Label()`, `format.uuid()`, `format.uri()`, etc.) with `validate()` returning `optional`
- Base64 encode/decode, jsonpatch.escapeKey (RFC 6901)
- `list.first()`, `list.last()`, `list.sort()`, `lists.range()`, `string.reverse()`, `CIDR.ip()`, `IP/CIDR.string()`, IP version helpers
- Compile `x-kubernetes-validations` rules into `cel::Program`
- Walk OpenAPI schema trees (properties, items, additionalProperties)
- Evaluate rules with `self`/`oldSelf` bindings, collect `ValidationError`s with field paths
- `messageExpression`, `optionalOldSelf`, pre-compiled schema reuse (`CompiledSchema`)
- Schema-aware `format: date-time` / `format: duration` support (auto-converts JSON strings to CEL `Timestamp`/`Duration`)
- Field name escaping — CEL reserved words + special chars auto-escaped to match K8s Go apiserver behavior
- `ErrorKind` classification, `fieldPath` override support

```rust
use kube_cel::validation::Validator;

let schema = &serde_json::to_value(&crd)?["spec"]["versions"][0]["schema"]["openAPIV3Schema"];
let object = serde_json::to_value(&my_resource)?;
let errors = Validator::new().validate(schema, &object, None);
```

kube-cel's internal `Rule` type and `kube-core::Rule` serialize to identical JSON, so they're already compatible at the `serde_json::Value` level.

Remaining gaps that need upstream `cel` crate support: `cel.bind()`, `list.sortBy()`, TwoVarComprehensions (K8s 1.33+). Authz library is out of scope (requires API server connection).

For tighter integration, kube-core could either depend on kube-cel behind a feature flag (like `tame-oauth` or `json-patch`), or implement the validation directly using the `cel` crate based on what kube-cel has built. Happy to contribute either way.

**Comment by doxxx93** [comment]
**Follow-up for Estimating the scope,**
                                                                                                                                                                                                                                                                           
After building the PoC, I tried to estimate the scope of work needed to cover K8s CEL validation end-to-end. 

It turned out significantly larger than I initially expected — full coverage of the Kubernetes CEL extension functions (strings, lists, sets, IP/CIDR, URLs, semver, quantity, math, format validators, etc.) adds up to ~8500 lines. But from what I can tell, that's the unavoidable cost if we want to faithfully support what the API server evaluates.

The PoC works and passes validation against the rules I've tested. There are areas to improve in the current implementation, but those would be covered before or during integration. 

> One thing I haven't explored yet is whether dropping the cel crate dependency entirely and building evaluation from scratch would be worth the effort — haven't measured the ROI there.

Here's how the work breaks down and could be split(approximate line counts for each unit of work)

> FYI) Test cases are largely ported from the upstream K8s CEL Go test suites.

1. Core engine + most-used functions (strings, lists, sets) — ~4100 lines (~50% tests). The minimum viable feature; rules like `self.replicas >= 0` or `self.name.startsWith('app-')` work with just this.
2. Additional function libraries, addable incrementally in any order. 
    - quantity — ~700 lines (~40% tests) `A`
    - semver — ~300 lines (~60% tests) `A`
    - math — ~500 lines (~50% tests)
    - IP/CIDR — ~700 lines (~50% tests)
    - URLs — ~300 lines (~50% tests)
    - regex — ~200 lines (~60% tests)
    - format — ~500 lines (~40% tests) `B`
    - named-format — ~1000 lines (~30% tests) `B`
    - encoders — ~100 lines (~60% tests)
    - jsonpatch — ~100 lines (~70% tests)

> Note that `A`s - `quantity`+`semver` share dispatch for `isGreaterThan`/`compareTo`, and `B` - `named-format` depends on `format` — these pairs would be better if landed together.
3. Derive + Admission integration — `validation_schema()` extraction from CRD, AdmissionResponse helpers — ~700 lines (~40% tests)

Happy to discuss further if this is something worth moving forward on — open to any direction on how to approach it.

**Comment by clux** [comment]
This is amazing. The crate works pretty much exactly as how I had hoped, and you've mapped out tons of these. Lots of code, but from what I can tell, it's feels like "write-once; unlikely to need many updates" type of code given it closely follows clear requirements.

I think a kube-core integration would be really nice. You could possibly even hook it into `CustomResourceExt` (implemented by kube-derive) in such a way that users could run a `Foo::validate` fn against an instance of a custom resource Foo without needing to think much about schemas/validator deps at all. But maybe there's an easier way to allow validating sub-structs also without needing to create a full instance to verify validation rules (i could see that being ergonomic in unit tests).

**Comment by doxxx93** [comment]
Thanks! Good to hear the direction aligns.

I have a rough breakdown ready (core engine, function libraries, derive + admission hooks) for the kube-core integration. Wanted to check — should I use sub-issues off this one to track each piece, or just a series of PRs referencing #1670? Also, do you prefer smaller incremental PRs or fewer larger ones?

For the `Foo::validate` / sub-struct validation via derive — I think that fits naturally as a follow-up once the core integration lands.

**Comment by clux** [comment]
Feel free to just track this issue for it for now.

Smaller incremental PRs is usually the easiest though (depending on how cleanly it breaks down).

**Comment by doxxx93** [comment]
#1954 is approved (pending merge) — that covers the core integration (re-export from `kube-cel`).

For the next step, I have a working prototype for derive integration. Here's how it works:

```rust
#[derive(CustomResource, Serialize, Deserialize, Clone, KubeSchema)]
#[kube(
    group = "example.com", version = "v1", kind = "Foo",
    namespaced,
    cel,                                    // ← enables validate_cel() generation
    validation = "self.spec.replicas >= 0"  // ← existing attribute, reused as-is
)]
struct FooSpec {
    replicas: i32,
}

// Generated methods on the root CR struct:
let foo = Foo::new("test", FooSpec { replicas: 3 });
let errors = foo.validate_cel();                     // create validation
let errors = new_foo.validate_cel_update(&old_foo);  // transition rules

// Sub-struct validation also works:
let errors = FooSpec::validate_cel(&spec_value, None);

// Compiled schema is cached via LazyLock — zero cost after first call.
```

Key decisions:
- **`#[kube(cel)]`** flag on the existing attribute — no new attribute namespace needed
- When `cel` is set, `KubeSchema` is auto-derived so `#[kube(validation = "...")]` rules are processed into the CRD schema
- Schema compiled from `Foo::crd()` output (root) / `schemars::JsonSchema` (sub-struct), cached in `LazyLock`
- Both root CR (`Foo::validate_cel()`) and spec struct (`FooSpec::validate_cel()`) get generated methods — the sub-struct version extracts schema via `JsonSchema` trait, so it's useful for unit testing without constructing a full CR

The derive changes for both are small (~30 lines each in `custom_resource.rs` and `cel_schema.rs`), so I'd include them in a single PR.

Does this direction make sense? Any preferences on the API shape before I open a PR for the derive work?

**Comment by clux** [comment]
I think this makes sense. Happy to merge your first one also.

> When cel is set, KubeSchema is auto-derived so #[kube(validation = "...")] rules are processed into the CRD schema

does this mean that `kube(cel)` is going to be mutually exclusive with any `kube(schema = "non-default-mode")? if so how would it interact with users importing a manual schema containing cel validations?

> Both root CR (Foo::validate_cel()) and spec struct (FooSpec::validate_cel()) get generated methods [..] useful for unit testing without constructing a full CR

👍. does this apply to any other substructs down the tree? i feel this could be useful also.

**Comment by doxxx93** [comment]
> does this mean that `kube(cel)` is going to be mutually exclusive with any `kube(schema = "non-default-mode")`?

No — `#[kube(cel)]` only forces `KubeSchema` when `schema_mode.derive()` is true (the default). With `schema = "Manual"`, the `cel` flag doesn't touch schema generation at all — `validate_cel()` is still generated and pulls the schema from `Foo::crd()` output, so manual schemas with CEL rules work as-is.

Verified by tracing the derive codegen — the relevant guard is:
```rust
// custom_resource.rs
if schema_mode.derive() && (!validations.is_empty() || cel_validate) {
    derive_paths.push(syn::parse_quote! { #kube::KubeSchema });
}
```
When `schema_mode` is `Manual`, `schema_mode.derive()` returns false, so `KubeSchema` is never forced. The `validate_cel()` impl block is generated unconditionally (gated only by the `cel` flag), and it reads the schema from `Foo::crd()` at runtime — so whatever schema the user provides (derived or manual) is what gets compiled and validated against.

> does this apply to any other substructs down the tree?

Yes — any struct deriving `KubeSchema` with `#[x_kube(cel)]` gets its own `validate_cel()`, regardless of nesting depth:

```rust
#[derive(KubeSchema, Serialize, Deserialize)]
#[x_kube(cel, validation = "self.timeout > 0")]
struct RetryConfig {
    timeout: i32,
}

// works at any level:
RetryConfig::validate_cel(&value, None);
```

Each struct opts in individually — the schema is extracted via `JsonSchema` trait and compiled independently per struct (cached in `OnceLock`). No depth limit.

Verified with integration tests — added `#[x_kube(cel)]` to a nested spec struct and confirmed `validate_cel()` is generated and works:
```rust
// CelNestedSpec has #[x_kube(cel, validation = "self.name.size() > 0")]
// and field-level #[x_kube(validation = "self.timeoutSeconds > 0")]

#[test]
fn cel_substruct_validate_passes() {
    let spec = serde_json::to_value(CelNestedSpec { name: "hello".into(), ... }).unwrap();
    let errors = CelNestedSpec::validate_cel(&spec, None);
    assert!(errors.is_empty());
}

#[test]
fn cel_substruct_validate_fails() {
    let spec = serde_json::to_value(CelNestedSpec { name: "".into(), ... }).unwrap(); // violates rules
    let errors = CelNestedSpec::validate_cel(&spec, None);
    assert!(errors.len() >= 2); // name + timeout violations
}
```
Both pass — 28/28 tests in `crd_schema_test` green.

**Comment by clux** [comment]
Amazing 👍

**Comment by doxxx93** [comment]
#1954 is ready to merge — happy to merge whenever. Will follow up with the derive integration (`#[kube(cel)]` + sub-struct `#[x_kube(cel)]`) as a separate PR.

**Comment by clux** [comment]
> Will follow up with the derive integration (`#[kube(cel)]` + sub-struct `#[x_kube(cel)]`) as a separate PR.

In case this got left to the wayside; you had a plan to do a;

> PR2 (feat/cel-derive): #[kube(cel)] derive macro generating Foo::validate_cel() methods

do you want to get that in?

**Comment by clux** [comment]
@doxxx93 I am writing up a headline part for this for 4.0.

I have the following paragraph below for now.  But maybe we should add an example, or info on how it hooks into custom resources. If you have preferences I can add it.

## CEL Validation
A new crate [kube-cel](https://docs.rs/kube-cel/latest/kube_cel/) is being re-exported through `kube::core::cel`.

Kubernetes CRDs support [CEL validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) via `x-kubernetes-validations`, but these are only evaluated server-side by the API server. This crate allows evaluating these rules locally using rules matching the upstream Kubernetes CEL libraries.

This is available with the new (non-default) `kube/cel` feature.

**Comment by doxxx93** [comment]
Yes, let's get this into 4.0. The March prototype is validated (root `#[kube(cel)]` and sub-struct `#[x_kube(cel)]`, 28/28 green at the time), but it predates the 0.6 flatten, so I'll rebase it onto current main and re-point the generated code at the flat `kube::core::cel` surface. Scope stays as planned: root + sub-struct in one PR (~30 lines each in `custom_resource.rs` / `cel_schema.rs`), plus doctests, `crd_schema_test` coverage, and a runnable `examples/` file the release notes can link to. I'll open it shortly.

---

Looks good. The "how it hooks into custom resources" part is the derive integration from #1670, which I'm landing in 4.0, so the headline can show the ergonomic form:

```
    #[derive(CustomResource, Serialize, Deserialize, Clone, KubeSchema)]
    #[kube(group = "example.com", version = "v1", kind = "Foo", namespaced, cel, validation = "self.spec.replicas >= 0")]
    struct FooSpec { replicas: i32 }

    let foo = Foo::new("test", FooSpec { replicas: -1 });
    let errors = foo.validate_cel();                     // creation rules
    let errors = new_foo.validate_cel_update(&old_foo);  // transition rules
```

I'll add a runnable example under examples/ in that PR so the note can link to it (same as how the garde section links to crd_api.rs).
